README.md
libqpsd
=======

PSD (Photoshop Document) & PSB (Photoshop Big) Plugin for Qt/C++ (Qt4/Qt5)


This project is based on the Photoshop File Format Specification found in http://www.adobe.com/devnet-apps/photoshop/fileformatashtml/PhotoshopFileFormats.htm

Format:
- [X] PSD
- [X] PSB
	
Compression:
- [X] Raw
- [X] RLE
- [ ] Zip without Prediction
- [ ] Zip with Prediction
	
Color Mode
- [X] Bitmap
- [X] Grayscale
- [X] Indexed
- [X] RGB
- [X] CMYK
- [X] Multichannel
- [X] Duotone
- [X] Lab
	
Depth
- [X] 1-bit (Bitmap)
- [X] 8-bits
- [X] 16-bits
- [ ] 32-bits (Tonemapping Algorithm, needed help!)
- *16-bit depth is scaled down to 8-bits*


CONTRIBUTE

- If you have an existing PSD/PSB file with different color mode or compression. Feel free to send me an [email](mailto:ronmarti18@gmail.com). I need samples with zip-compressed layers (zip w/ or w/o prediction).
- If you want to contribute to the code, just fork the project and pull requests.
- If you want to keep this project alive you can send your donations to:
	- Bitcoin: 3A23hHJF8q8hNPz3sedqD9T7g25ELPcvg2
	- Paypal: https://www.paypal.me/RonieMartinez

CONTRIBUTORS

- Ronie Martinez aka Code ReaQtor (ronmarti18@gmail.com)
- Yuezhao Huang (huangezhao@gmail.com)
- username "asgohtals" (http://qt-project.org/member/136052) - contributed PSB files for testing: 
	- berlin-cmyk.psb
	- berlin-quadtone.psb
	- wall-small-Lab.psb

LIST OF SOFTWARE USING libqpsd
- [nomacs](http://nomacs.org/)
- [EzViewer](https://github.com/yuezhao/ezviewer)
- [PhotoQt](http://photoqt.org/)
- [Seer](http://www.1218.io/)
- *for applications not included in the list, shoot me an email*


CHANGELOG.md
List of changes to libqpsd

## [Unreleased]
## [v2.0.1] - 2018-11-07
### Changed
- Version for release (#23)
- Update README
- Update copyright

### Fixes
- QImage format on Qt 5.10+ (#19)
- 'make install' (#21)
- Crash for PSB files greater than 2GB (#22)

## [v2.0.0] - 2017-10-31
### Changed
- Updated copyright date
- Refactored code

[Unreleased]: https://github.com/Code-ReaQtor/libqpsd/compare/v2.0.1...master 
[v2.0.1]: https://github.com/Code-ReaQtor/libqpsd/releases/tag/v2.0.1
[v2.0.0]: https://github.com/Code-ReaQtor/libqpsd/releases/tag/v2.0.0


qpsdplugin.h
/*
Copyright (c) 2012-2018 Ronie Martinez (ronmarti18@gmail.com)
All rights reserved.

This library is free software; you can redistribute it and/or
modify it under the terms of the GNU Lesser General Public
License as published by the Free Software Foundation; either
version 2.1 of the License, or (at your option) any later version.

This library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied
warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
PURPOSE.  See the GNU Lesser General Public License for more
details.

You should have received a copy of the GNU Lesser General Public
License along with this library; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
02110-1301  USA
*/

#ifndef QPSDPLUGIN_H
#define QPSDPLUGIN_H

#include <QImageIOPlugin>


class QPsdPlugin : public QImageIOPlugin
{
    Q_OBJECT
#if QT_VERSION >= 0x050000
    Q_PLUGIN_METADATA(IID "org.qt-project.Qt.QImageIOHandlerFactoryInterface" FILE "QPsdPlugin.json")
#endif // QT_VERSION >= 0x050000
    
public:
    QPsdPlugin(QObject *parent = 0);
    ~QPsdPlugin();
    QStringList keys() const;
    Capabilities capabilities(QIODevice *device,
                    const QByteArray &format) const;
    QImageIOHandler *create(QIODevice *device,
                               const QByteArray &format = QByteArray()) const;
};

#endif // QPSDPLUGIN_H


qpsdhandler.h
/*
Copyright (c) 2012-2018 Ronie Martinez (ronmarti18@gmail.com)
All rights reserved.

This library is free software; you can redistribute it and/or
modify it under the terms of the GNU Lesser General Public
License as published by the Free Software Foundation; either
version 2.1 of the License, or (at your option) any later version.

This library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied
warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
PURPOSE.  See the GNU Lesser General Public License for more
details.

You should have received a copy of the GNU Lesser General Public
License along with this library; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
02110-1301  USA
*/

#ifndef QPSDHANDLER_H
#define QPSDHANDLER_H

#include <QImageIOHandler>
#include <QImage>
#include <QColor>
#include <QVariant>
#include <qmath.h>

class QPsdHandler : public QImageIOHandler
{
public:
    QPsdHandler();
    ~QPsdHandler();

    bool canRead() const;
    bool read(QImage *image);
    //bool write(const QImage &image);

    static bool canRead(QIODevice *device);

    QVariant option(ImageOption option) const;
    //void setOption(ImageOption option, const QVariant &value);
    bool supportsOption(ImageOption option) const;
private:
    bool isValidSignature(quint32 signature);
    bool isValidVersion(quint16 version);
    bool isChannelCountSupported(quint16 channel);
    bool isValidWidthOrHeight(quint16 version, quint32 value);
    bool isSupportedDepth(quint16 depth);
    bool isSupportedColorMode(quint16 colorMode);
    QByteArray readColorData(QDataStream& input);
    void skipImageResources(QDataStream& input);
    void skipLayerAndMaskSection(QDataStream& input);
    enum Compression {
        RAW = 0,
        RLE = 1,
        ZIP_WITHOUT_PREDICTION = 2,
        ZIP_WITH_PREDICTION = 3
    };
    QByteArray readImageData(QDataStream& input, Compression compression, quint64 size=0);
    enum ColorMode {
        BITMAP = 0,
        GRAYSCALE = 1,
        INDEXED = 2,
        RGB = 3,
        CMYK = 4,
        MULTICHANNEL = 7,
        DUOTONE = 8,
        LAB = 9,
    };
    QImage processBitmap(QByteArray& imageData, quint32 width, quint32 height);
    QImage processGrayscale8(QByteArray& imageData, quint32 width, quint32 height);
    QImage processGrayscale8WithAlpha(QByteArray& imageData, quint32 width, quint32 height,
                                     quint64 totalBytesPerChannel);
    QImage processGrayscale16(QByteArray& imageData, quint32 width, quint32 height);
    QImage processGrayscale16WithAlpha(QByteArray& imageData, quint32 width, quint32 height,
                                       quint64 totalBytesPerChannel);
    QImage processIndexed(QByteArray& colorData, QByteArray& imageData, quint32 width,
                          quint32 height);
    QImage processRGB8(QByteArray& imageData, quint32 width, quint32 height,
                       quint64 totalBytesPerChannel);
    QImage processRGB16(QByteArray& imageData, quint32 width, quint32 height,
                        quint64 totalBytesPerChannel);
    QImage processRGB8WithAlpha(QByteArray& imageData, quint32 width, quint32 height,
                                quint64 totalBytesPerChannel);
    QImage processRGB16WithAlpha(QByteArray& imageData, quint32 width, quint32 height,
                                 quint64 totalBytesPerChannel);
    QImage processCMY8(QByteArray& imageData, quint32 width, quint32 height,
                       quint64 totalBytesPerChannel);
    QImage processCMYK8(QByteArray& imageData, quint32 width, quint32 height,
                        quint64 totalBytesPerChannel);
    QImage processCMYK8WithAlpha(QByteArray& imageData, quint32 width, quint32 height,
                                 quint64 totalBytesPerChannel);
    QImage processCMYK16(QByteArray& imageData, quint32 width, quint32 height,
                         quint64 totalBytesPerChannel);
    QImage processCMYK16WithAlpha(QByteArray& imageData, quint32 width, quint32 height,
                                  quint64 totalBytesPerChannel);
    QImage processDuotone(QByteArray& imageData, quint32 width, quint32 height);
    QImage processLAB8(QByteArray& imageData, quint32 width, quint32 height,
                       quint64 totalBytesPerChannel);
    QImage processLAB8WithAlpha(QByteArray& imageData, quint32 width, quint32 height,
                                quint64 totalBytesPerChannel);
    QImage processLAB16(QByteArray& imageData, quint32 width, quint32 height,
                        quint64 totalBytesPerChannel);
    QImage processLAB16WithAlpha(QByteArray& imageData, quint32 width, quint32 height,
                                 quint64 totalBytesPerChannel);
};

#endif // QPSDHANDLER_H


qpsdplugin.cpp
/*
Copyright (c) 2012-2018 Ronie Martinez (ronmarti18@gmail.com)
All rights reserved.

This library is free software; you can redistribute it and/or
modify it under the terms of the GNU Lesser General Public
License as published by the Free Software Foundation; either
version 2.1 of the License, or (at your option) any later version.

This library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied
warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
PURPOSE.  See the GNU Lesser General Public License for more
details.

You should have received a copy of the GNU Lesser General Public
License along with this library; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
02110-1301  USA
*/

#include "qpsdplugin.h"
#include "qpsdhandler.h"

QPsdPlugin::QPsdPlugin(QObject *parent) :
    QImageIOPlugin(parent)
{
}

QPsdPlugin::~QPsdPlugin()
{
}

QStringList QPsdPlugin::keys() const
{
    return QStringList() << "psd" << "psb";
}

QImageIOPlugin::Capabilities QPsdPlugin::capabilities(
    QIODevice *device, const QByteArray &format) const
{
    if (format == "psd" || format == "psb")
        return Capabilities(CanRead);//TODO: add CanWrite support

    if (!(format.isEmpty() && device->isOpen()))
        return 0;

    Capabilities cap;
    if (device->isReadable() && QPsdHandler::canRead(device))
        cap |= CanRead;
    return cap;
}

QImageIOHandler *QPsdPlugin::create(
    QIODevice *device, const QByteArray &format) const
{
    QImageIOHandler *handler = new QPsdHandler;
    handler->setDevice(device);
    handler->setFormat(format);
    return handler;
}

#if QT_VERSION < 0x050000
Q_EXPORT_STATIC_PLUGIN(QPsdPlugin)
Q_EXPORT_PLUGIN2(QPsd, QPsdPlugin)
#endif // QT_VERSION < 0x050000


qpsdhandler.cpp
/*
Copyright (c) 2012-2018 Ronie Martinez (ronmarti18@gmail.com)
All rights reserved.

This library is free software; you can redistribute it and/or
modify it under the terms of the GNU Lesser General Public
License as published by the Free Software Foundation; either
version 2.1 of the License, or (at your option) any later version.

This library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied
warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
PURPOSE.  See the GNU Lesser General Public License for more
details.

You should have received a copy of the GNU Lesser General Public
License along with this library; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
02110-1301  USA
*/

#include "qpsdhandler.h"
#ifdef QT_DEBUG
#include <QDebug>
#include <QElapsedTimer>
#endif


QPsdHandler::QPsdHandler()
{
}

QPsdHandler::~QPsdHandler()
{
}

bool QPsdHandler::canRead() const
{
    if (canRead(device())) {
        QByteArray bytes = device()->peek(6);
        QDataStream input(bytes);
        input.setByteOrder(QDataStream::BigEndian);
        quint32 signature;
        quint16 version;
        input >> signature >> version;
        if (version == 1)
            setFormat("psd");
        else if (version == 2)
            setFormat("psb");
        else return false;
        return true;
    }
    return false;
}

bool QPsdHandler::read(QImage *image)
{
    QDataStream input(device());
    quint32 signature, height, width;
    quint16 version, channels, depth, colorMode, compression;

    input.setByteOrder(QDataStream::BigEndian);

    input >> signature;
    if (!isValidSignature(signature)) {
        return false;
    }

    input >> version;
    if (!isValidVersion(version)) {
        return false;
    }

    input.skipRawData(6); //reserved bytes should be 6-byte in size

    input >> channels;
    if (!isChannelCountSupported(channels)) {
        return false;
    }

    input >> height;
    if (!isValidWidthOrHeight(version, height)) {
        return false;
    }

    input >> width;
    if (!isValidWidthOrHeight(version, width)) {
        return false;
    }

    input >> depth;
    if (!isSupportedDepth(depth)) {
        return false;
    }

    input >> colorMode;
    if (!isSupportedColorMode(colorMode)) {
        return false;
    }

    QByteArray colorData = readColorData(input);
    skipImageResources(input);
    skipLayerAndMaskSection(input);

    input >> compression;

    quint64 totalBytesPerChannel = (quint64)width * height * depth / 8;
    quint64 size = (quint64)channels * totalBytesPerChannel;
    QByteArray imageData;

    switch (compression) {
    case RLE:
//        The RLE-compressed data is preceeded by a 2-byte(psd) or 4-byte(psb)
//        data count for each row in the data
        if (format() == "psd")
            input.skipRawData(height * channels * 2);
        else if (format() == "psb")
            input.skipRawData(height * channels * 4);
    case RAW:
        imageData = readImageData(input, (Compression)compression, size);
        break;
    default:
        return false;
        break;

    }

    if (input.status() != QDataStream::Ok)
        return false;

    if ((quint64)imageData.size() != size)
        return false;

#ifdef QT_DEBUG
    qDebug() << "format: " << format()
             << "\ncolor mode: " << colorMode
             << "\ndepth: " << depth
             << "\nchannels: " << channels
             << "\ncompression: " << compression
             << "\nwidth: " << width
             << "\nheight: " << height
             << "\ntotalBytesPerChannel: " << totalBytesPerChannel
             << "\nimage data: " << imageData.size();
#endif


    switch (colorMode) {
    case BITMAP:
    {
        QImage result = processBitmap(imageData, width, height);
        if (result.isNull())
            return false;
        else {
            *image = result;
            return true;
        }
    }
        break;
    case GRAYSCALE:
    {
        switch (depth) {
        case 8:
            if (1 == channels) {
                *image = processGrayscale8(imageData, width, height);
                return true;
            } else {
                //excess channels other than Gray are considered alphas
                *image = processGrayscale8WithAlpha(imageData, width, height, totalBytesPerChannel);
                return true;
            }
            break;
        case 16:
            if (1 == channels) {
                *image = processGrayscale16(imageData, width, height);
                return true;
            } else {
                //excess channels other than Gray are considered alphas
                *image = processGrayscale16WithAlpha(imageData, width, height, totalBytesPerChannel);
                return true;
            }
            break;
        case 32: // FIXME: 32 bpc (HDR)... requires tonemapping
        default:
            return false;
            break;
        }
    }
        break;
    case INDEXED:
        if (8 == depth && 1 == channels) {
            *image = processIndexed(colorData, imageData, width, height);
            return true;
        } else {
            return false;
        }
        break;
    case RGB:
    {
        switch (depth) {
        case 8:
            if (3 == channels) {
                *image = processRGB8(imageData, width, height, totalBytesPerChannel);
                return true;
            } else {
                //excess channels other than RGB are considered alphas
                *image = processRGB8WithAlpha(imageData, width, height, totalBytesPerChannel);
                return true;
            }
            break;
        case 16:
            if (3 == channels) {
                *image = processRGB16(imageData, width, height, totalBytesPerChannel);
                return true;
            } else {
                //excess channels other than RGB are considered alphas
                *image = processRGB16WithAlpha(imageData, width, height, totalBytesPerChannel);
                return true;
            }
            break;
        case 32: //FIXME: 32 bpc (HDR)... requires tonemapping
        default:
            return false;
            break;
        }
    }
        break;
    /* Mixed CMYK and Multichannel logic due to similarities*/
    case CMYK:
    case MULTICHANNEL:
    {
//        Reference: http://help.adobe.com/en_US/photoshop/cs/using/WSfd1234e1c4b69f30ea53e41001031ab64-73eea.html#WSfd1234e1c4b69f30ea53e41001031ab64-73e5a
//        Converting a CMYK image to Multichannel mode creates cyan, magenta, yellow, and black spot channels.
//        Converting an RGB image to Multichannel mode creates cyan, magenta, and yellow spot channels.
        switch (depth) {
        case 8:
            if (3 == channels) {
                *image = processCMY8(imageData, width, height, totalBytesPerChannel);
                return true;
            } else if (4 == channels) {
                *image = processCMYK8(imageData, width, height, totalBytesPerChannel);
                return true;
            } else {
                *image = processCMYK8WithAlpha(imageData, width, height, totalBytesPerChannel);
                return true;
            }
            break;
        case 16:
            if (4 == channels) {
                *image = processCMYK16(imageData, width, height, totalBytesPerChannel);
                return true;
            } else {
                *image = processCMYK16WithAlpha(imageData, width, height, totalBytesPerChannel);
                return true;
            }
            break;
        default:
            return false;
            break;
        }
    }
        break;
    case DUOTONE:
//        Duotone images: color data contains the duotone specification
//        (the format of which is not documented). Other applications that
//        read Photoshop files can treat a duotone image as a gray image,
//        and just preserve the contents of the duotone information when
//        reading and writing the file.
//        TODO: find a way to actually get the duotone, tritone, and quadtone colors
        if (8 == depth && 1 == channels) {
            *image = processDuotone(imageData, width, height);
            return true;
        } else {
            return false;
        }
        break;
    case LAB:
    {
        switch (depth) {
        case 8:
            if (3 == channels) {
                *image = processLAB8(imageData, width, height, totalBytesPerChannel);
                return true;
            } else {
                *image = processLAB8WithAlpha(imageData, width, height, totalBytesPerChannel);
                return true;
            }
            break;
        case 16:
            if (3 == channels) {
                *image = processLAB16(imageData, width, height, totalBytesPerChannel);
                return true;
            } else {
                *image = processLAB16WithAlpha(imageData, width, height, totalBytesPerChannel);
                return true;
            }
            break;
        default:
            return false;
            break;
        }
    }
        break;
    default:
        return false;
        break;
    }
    return input.status() == QDataStream::Ok;
}

bool QPsdHandler::canRead(QIODevice *device)
{
    return device->peek(4) == "8BPS";
}

QVariant QPsdHandler::option(ImageOption option) const
{
    if (option == Size) {
        QByteArray bytes = device()->peek(26);
        QDataStream input(bytes);
        quint32 signature, height, width;
        quint16 version, channels, depth, colorMode;
        input.setByteOrder(QDataStream::BigEndian);
        input >> signature >> version ;
        input.skipRawData(6); //reserved bytes should be 6-byte in size
        input >> channels >> height >> width >> depth >> colorMode;
        if (input.status() == QDataStream::Ok && signature == 0x38425053 &&
                (version == 1 || version == 2))
            return QSize(width, height);
    }
    return QVariant();
}

bool QPsdHandler::supportsOption(ImageOption option) const
{
    return option == Size;
}
